### PR TITLE
Init test beans lazily

### PR DESCRIPTION
**Pull request by rabiori**
-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [ ] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

Something we did in our project, which cut down our test-time by ~1 min - I though it might be helpful to other devs.

I did not create an issue yet, I wanted to see if this seems like a welcome addition or not.

### CODE REVIEW

**Comment by rabiori** [comment]
@jkutner @jdigger @gunnarahlberg @pascalgrimaud - sorry to tag you guys, but I think this PR got lost somehow?

**Comment by gmarziou** [comment]
Interesting, I'm sure JHipster team would love to cut down continuous integration time.
Few comments:

- documentation would be required (jhttps://www.jhipster.tech/profiles/#-profiles and project's README.md)
- where do you enable "eagerBeanInit" profile? In maven or gradle, in application properties?
- profile name (camel-cased) is quite different other JHipster profiles (kebab-cased). It should probably be renamed as "eager-bean-init"

File: generators/server/templates/src/test/java/package/config/TestLazyBeanInitConfiguration.java.ejs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * <br/>
+ * In order to override this for any specific test, add to it: {@code @ActiveProfiles(TestLazyBeanInitConfiguration.EAGER_BEAN_INIT)}
+ * <br/>
+ * When working with spring-boot 2.2.0 or later, should be replaced with property: <code>spring.main.lazy-initialization</code>.
+ * <br/>
+ * See <a href="https://spring.io/blog/2019/03/14/lazy-initialization-in-spring-boot-2-2">lazy-initialization-in-spring-boot-2-2</a>.
+ */
+@Component
+@Profile("!" + TestLazyBeanInitConfiguration.EAGER_BEAN_INIT)
+public class TestLazyBeanInitConfiguration {
```
Comment (vishal423) [inline generators/server/templates/src/test/java/package/config/TestLazyBeanInitConfiguration.java.ejs]: shouldn't you need to implement `BeanFactoryPostProcessor`?

Comment (rabiori) [reply generators/server/templates/src/test/java/package/config/TestLazyBeanInitConfiguration.java.ejs]: you are completely right, probably got lost on the way to the PR, I'll return it

**Comment by rabiori** [comment]
> Interesting, I'm sure JHipster team would love to cut down continuous integration time.
> Few comments:
> 
> * documentation would be required (https://www.jhipster.tech/profiles/#-profiles and project's README.md)
> * where do you enable "eagerBeanInit" profile? In maven or gradle, in application properties?
> * profile name (camel-cased) is quite different from other JHipster profiles (kebab-cased). It should probably be renamed as "eager-bean-init"

In our project, in any test that we wanted to eagerly init the beans I added the following:
```
@ActiveProfiles(TestLazyBeanInitConfiguration.EAGER_BEAN_INIT)
```
You can pass it by gradle or maven, but the way I implemented this means that by default all beans will be lazy, and any specific test that requires otherwise can change it.

Regarding the profile name, I changed it as commented.

**Comment by gmarziou** [comment]
Great but I did not understand how and where you would document it?

**Comment by gmarziou** [comment]
I ran a little test with jhipster-sample-app-microservice app and 8 entities (https://github.com/jhipster/jdl-samples/blob/master/Oracle-Human-Resources-sample.jdl) `mwnw verify` ran in 43 s instead of 45 s.

**Comment by rabiori** [comment]
> I ran a little test with jhipster-sample-app-microservice app and 8 entities (https://github.com/jhipster/jdl-samples/blob/master/Oracle-Human-Resources-sample.jdl) `mwnw verify` ran in 43 s instead of 45 s.

Perhaps in our code we have many more modules.
I don't see the harm in merging this, but if you think this is superfluous then just let me know.

**Comment by gmarziou** [comment]
I don't think it's superfluous, I'm sure it helps when number of entities grows.

When JHipster upgrades to Spring Boot 2.2,  should we remove this class in favor of  setting `spring.main.lazy-initialization` property to true in `src/test/resources/application.yml` ?

**Comment by Blackdread** [comment]
Yes for `spring.main.lazy-initialization`

**Comment by rabiori** [comment]
> I don't think it's superfluous, I'm sure it helps when number of entities grows.
> 
> When JHipster upgrades to Spring Boot 2.2, should we remove this class in favor of setting `spring.main.lazy-initialization` property to true in `src/test/resources/application.yml` ?

indeed, and I added it to the javadoc of the class as well:
```
When working with spring-boot 2.2.0 or later, should be replaced with property: <code>spring.main.lazy-initialization</code>.
```

**Comment by gmarziou** [comment]
> indeed, and I added it to the javadoc of the class as well
Yes this is where I read it. :)

So I'm wondering whether it makes sense to add this class if we are going to remove it as soon as we upgrade to SB 2.2. which should be soon I suppose. @pascalgrimaud @jdubois @deepu105 what do you think?

**Comment by murdos** [comment]
> 
> So I'm wondering whether it makes sense to add this class if we are going to remove it as soon as we upgrade to SB 2.2. which should be soon I suppose. @pascalgrimaud @jdubois @deepu105 what do you think?

This is exactly what I was thinking. IMO we should just wait for SB 2.2 and don't include something that we know we'll remove soon.

**Comment by DanielFran** [comment]
+1 to wait SB 2.2

**Comment by vishal423** [comment]
I checked few builds(~14 entities) with and without this change to see actual performance gain, however, couldn't see any specific gain.
### Build including this change:
https://dev.azure.com/jhipster/generator-jhipster/_build/results?buildId=5116&view=logs
#### Backend tests took around `3m 21s`

### Build before this change:
https://dev.azure.com/jhipster/generator-jhipster/_build/results?buildId=5115
#### Backend tests took around `3m`

**Comment by jdubois** [comment]
I would rather wait for Spring Boot 2.2 : I did the same trick earlier, but for running my app (not just tests!) and I had several annoying corner cases, which made me lose a lot of time.
So if we push this to people, I'm afraid they might have the same issues: using the "official" Spring Boot 2.2 option should make this easier to troubleshoot. Then, I haven't checked how the Spring team does it, and if they indeed take into consideration those corner cases (basically you need to eager-init some beans, so you need a specific config for that).

**Comment by rabiori** [comment]
> I would rather wait for Spring Boot 2.2 : I did the same trick earlier, but for running my app (not just tests!) and I had several annoying corner cases, which made me lose a lot of time.
> So if we push this to people, I'm afraid they might have the same issues: using the "official" Spring Boot 2.2 option should make this easier to troubleshoot. Then, I haven't checked how the Spring team does it, and if they indeed take into consideration those corner cases (basically you need to eager-init some beans, so you need a specific config for that).

That's exactly the reason I did this in the tests only, and configured an "out" for when someone needed it.

If you intend to upgrade to 2.2 soon (you have a ticket for that?), I wouldn't merge this.
I just implemented this as a temporary measure, since I wasn't sure when you intended to upgrade.

**Comment by atomfrede** [comment]
I think we could write a small [tip in our tips and tricks](https://www.jhipster.tech/tips/) section as end of month the first 2.2 release candidate should be released: https://twitter.com/snicoll/status/1173586725131739136
 So I would close this in favor of boot 2.2. @rabiori are you willing to provide a tip for all users that can't upgrade when 2.2 is released?

**Comment by atomfrede** [comment]
I have created a tip, so let's close this PR and see how boot 2.2. works out.

See  https://github.com/jhipster/jhipster.github.io/pull/810

**Comment by pascalgrimaud** [comment]
thanks for the proposal
I'm closing this as it goes to a tips rather than directly to the main generator-jhipster
